pkg/messaging/nats/pubsub.go

```diff
@@ -69,20 +70,31 @@ func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler)
 	if topic == "" {
 		return ErrEmptyTopic
 	}
+
 	ps.mu.Lock()
-	defer ps.mu.Unlock()
 	// Check topic
 	s, ok := ps.subscriptions[topic]
 	switch ok {
 	case true:
 		// Check topic ID
 		if _, ok := s[id]; ok {
-			return ErrAlreadySubscribed
+			// Unlocking, so that Unsubscribe() can access ps.subscriptions
+			ps.mu.Unlock()
+			if err := ps.Unsubscribe(id, topic); err != nil {
+				return err
+			}
+
+			ps.mu.Lock()
+			if _, ok = ps.subscriptions[topic]; !ok {
+				ps.subscriptions[topic] = make(map[string]subscription)
+			}
 		}
 	default:
 		s = make(map[string]subscription)
 		ps.subscriptions[topic] = s
 	}
+	defer ps.mu.Unlock()
```

Comment:
I am not sure if this `defer` should be put here or under previous `Lock()`, where it was before. Both are valid, and even if you are putting Unlock-Lock after, it is not confusing, because we know that second `Lock()` is here just to come back to previous state.

I let @dborovcanin decide what place is better for this `defer()`.

Reply:
This place is better. While it looks a little more dangerous, it's actually better because you do not have to call an artificial `lock` so that an earlier `defer` unlocks it.